### ISSUE TITLE
Wire the dedicated gemma4 forward onto the resident-quant path — f32-only makes the arch unservable at any real size

### ISSUE THREAD

**Issue by anthony-saronic** [anthony-chaudhary/fak#5495]
**Title:** `feat(model): wire the dedicated gemma4 forward onto the resident-quant path — f32-only makes the arch unservable at any real size`

**Labels:** `enhancement`, `model-arch`, `track/A-model-support`, `class:dev`

## The claim

`internal/model/arch_support_gemma4.go` ships a typed refusal — `ResidentQuantUnsupportedError` —
whose message ends: *"the dedicated gemma4 forward is not wired on the resident-quant path
(issue #4274). Serve this architecture on the f32 (non-quant) forward."*

That refusal was the right call for #4274: the alternative was a panic deep in the shared qk-norm
band, because gemma4 interleaves two attention regimes with **different per-layer `head_dim`** and
the generic band passes the scalar `cfg.HeadDim` to `applyQKNormCfg`. #4274 is closed, and the
refusal is what shipped.

But the capability gap it names is still open, and it is load-bearing: **f32-only means gemma4 is
supported only at sizes nobody serves.** A 26B-class checkpoint at f32 is ~100 GB of weights. On any
single card in the ≤48 GB class the arch is, in practice, unsupported — the dedicated forward exists
and cannot be reached by a real checkpoint.

## What "wired" has to mean

The refusal's own comment states the acceptance bar: *"real work that needs a correctness witness
(logit cosine vs a reference run) before it can be trusted."* That bar should stay.

## Definition of done

- [ ] The resident-quant CPU forward dispatches to the gemma4 path (`ForwardGemma4`,
      `arch_support.go:97`) instead of running the shared uniform-geometry block code.
- [ ] `applyQKNormCfg` receives the **per-layer** `head_dim` (`headDimForLayer`), not the scalar.
- [ ] Correctness witness: logit cosine of the resident-quant forward vs the f32 reference on the
      same prompt, recorded in the issue.
- [ ] `ResidentQuantUnsupportedError` is **kept** for geometries still genuinely unsupported — the
      fix is to shrink its domain, never to delete the refusal.
- [ ] A test asserts a gemma4-geometry resident-quant session now prefills instead of refusing, and
      that a still-unsupported geometry refuses by name.

**Comment by anthony-chaudhary** [comment]
## Correctness witness — recorded, and its denominator widened first

Four of the five DoD bullets were already implemented in the tree before tonight. The one that was not is the third: *"Correctness witness: logit cosine of the resident-quant forward vs the f32 reference on the same prompt, **recorded in the issue**."* This issue had zero comments, so that bullet was outstanding even though the test computing the number existed.

Before recording it I looked at what the number was actually measured over, and it was too thin to publish.

### The witness as it stood: a cosine over 3 values

`TestGemma4ResidentQuantSessionPrefillsOnDedicatedForward` (`internal/ggufload/gemma4_session_test.go`) scored:

```
gemma4 resident-Q8_0 session vs f32 reference:
  logit cosine = 1.000000000000 over 3 finite logits, max|delta| = 1.3828277587890625e-05
```

`tinyGemma4GGUF` declared a 5-token vocabulary, two of which the fixture suppresses to `-inf`. So the cosine was computed over **three** numbers. Two short vectors of similarly-scaled logits correlate above 0.999 whether or not the forward is right, so `cosine = 1.000000000000` there is close to uninformative — publishing it as the correctness witness for a kernel forward path would have overstated what was checked. It is the same denominator problem the grounding epic (#5601 / #5602) is about, showing up in a numeric witness instead of a gate.

### Widened, then measured

`803dc2c792` — `test(ggufload): widen the gemma4 correctness witness denominator from 3 logits to 62 (#5495) (fak ggufload)`, on `origin/main`, `dos commit-audit` witnessed.

The fixture's vocabulary is now a named constant `tinyGemma4Vocab = 64`. Widening it took two coupled edits, and the second is the one that matters: the loader reads `VocabSize` from the **length of `tokenizer.ggml.tokens`**, not from the embedding tensor's dims. Growing only the embedding left the LM head at 5 and the witness silently just as narrow — the first attempt did exactly that and the shape assertion caught it. Both now derive from the one constant, and the token list has a comment saying why it must track it.

Nothing else in the package depended on the old width; `go test ./internal/ggufload` is green (9.5s, full package).

### The recorded witness

```
gemma4 resident-Q8_0 session vs f32 reference:
  logit cosine = 0.999999999999 over 62 finite logits, max|delta| = 8.20159912109375e-05
```

Fixture: 4 layers (3 sliding at head_dim 16 / 2 kv heads, 1 global at head_dim 32 / 1 kv head with no `v_proj`), hidden 32, vocab 64 less the 2 suppressed ids, prompt `[0,1,2]`, resident-Q8_0 session vs the f32 `LoadModel` of the same checkpoint.

That number carries weight the previous one did not, because the same test also pins the three things a cosine alone can hide:

- **Greedy decode agrees** — `finiteArgmax(quant) == finiteArgmax(f32)`. A high cosine with a flipped argmax would be a real regression the cosine would not show.
- **The premise is lossy** — `max|delta| != 0` is asserted, so if the "quantized" model were secretly the f32 one the test fails as vacuous rather than reporting a free 1.0.
- **The bridge attends its history** — `Prefill(last token alone)` must differ from `Prefill(whole prompt)`, so a context-blind bridge cannot pass by scoring only the newest token.

### DoD status

- [x] Resident-quant CPU forward dispatches to the gemma4 path — `kv.go` routes `Prefill` / `PrefillNoLogits` / `Step` to `gemma4Ingest` / `gemma4StackLast` when `Cfg.isGemma4()`.
- [x] `applyQKNormCfg` receives the per-layer `head_dim` — `gemma4.go:116` takes `hd := cfg.headDimForLayer(l)` and passes it to both the q and k norm calls.
- [x] **Correctness witness recorded** — above.
- [x] `ResidentQuantUnsupportedError` kept, domain shrunk — it still fires for a heterogeneous-`head_dim` arch with no dedicated forward, and gained a `Mode` field so a gemma4 session on an unwired mode (device backend, Metal, dynamic precision) refuses by name instead of running the wrong geometry.
- [x] Tests both ways — `TestGemma4ResidentQuantSessionPrefillsOnDedicatedForward` (prefills instead of refusing) and `TestGemma4StillUnsupportedSessionModesRefuseByName` (still-unsupported geometry refuses by name).

All five now hold. I have left this **open** rather than closing it, since the implementation landed under other commits and their author should get the close.

### One caveat worth stating plainly

This witness runs on a synthetic 4-layer fixture, not a real 26B-class Gemma-4 checkpoint. It proves the dedicated forward and the resident-quant store agree on the *geometry and the instruction stream*; it does not prove a production checkpoint decodes sensibly, and it does not touch the memory claim in the issue's framing (that f32-only makes the arch unservable at real sizes). A real-checkpoint decode remains worth capturing when one is staged — that is a hardware/weights-gated bullet, not something this host can produce, and I am not claiming it.

**Comment by anthony-chaudhary** [comment]
Independently reconciled by the guarded fleet closure audit: origin/main contains issue-bound resolving commit 803dc2c7921d ($(@{Issue=5495; Reason=; SHA=803dc2c7921d3cf2010f0b4ceea74a976d0388fe; Subject=test(ggufload): widen the gemma4 correctness witness denominator from 3 logits to 62 (#5495) (fak ggufload); Title=Wire the dedicated gemma4 forward onto the resident-quant path — f32-only makes the arch unservable at any real size}.Subject)), while this tracker remained OPEN and was not explicitly REOPENED. Closing stale issue state against the trunk witness.
